### PR TITLE
More configure fixes

### PR DESCRIPTION
**Pull request by MisterDA**
This PR attempts to remove accumulated Autoconf cruft, outdated code/comments, and fix bugs.

Unfortunately the commit `configure: set the value of AC_DEFINE macros to 1` adds some noise to this PR.

No change entry needed.
cc @shindere

### CODE REVIEW

**Review by ghost — APPROVED** [review]
LGTM.

I am moderately worried about the macOS specific changes, though - regarding the detection of `getentropy`, maybe older versions actually needed `<sys/random.h>` at some point, have you been able to check this?

**Comment by MisterDA** [comment]
> I am moderately worried about the macOS specific changes, though - regarding the detection of `getentropy`, maybe older versions actually needed `<sys/random.h>` at some point, have you been able to check this?

The [manual](https://opensource.apple.com/source/xnu/xnu-7195.81.3/bsd/man/man2/getentropy.2.auto.html) states it was introduced in Mac OS X 10.12 (2016). I have no proof, but it's likely that it was never made available in another header.
If you're concerned about removing the header inclusion in the autoconf test, note that it doesn't change anything: the compiler may warn that it cannot find a previous declaration, but will still link the test program, and autoconf will discover `getentropy(2)`.
Am I understanding your concern correctly? Does this answer it?

**Comment by ghost** [comment]
> The [manual](https://opensource.apple.com/source/xnu/xnu-7195.81.3/bsd/man/man2/getentropy.2.auto.html) states it was introduced in Mac OS X 10.12 (2016). I have no proof, but it's likely that it was never made available in another header. If you're concerned about removing the header inclusion in the autoconf test, note that it doesn't change anything: the compiler may warn that it cannot find a previous declaration, but will still link the test program, and autoconf will discover `getentropy(2)`. Am I understanding your concern correctly? Does this answer it?

I had misread the changes to `runtime/sys.c` and wrongly had the impression that the inclusion of `<sys/random.h>` had been removed. For the `autoconf` tests it does indeed not matter to explicitly include headers when only a linking test is performed.

**Review by dra27 — APPROVED** [review]
I like every aspect of this PR, thank you 🙂 Adding a green tick, but I'll let @shindere have a chance to comment before merging.

**Comment by shindere** [comment]
I second @dra27 in finding every aspect of this work excellent.

Thanks a lot, @MisterDA.

Re:the commit entitled "Keep former macros for compatibility", two
questions:

1. Wouldn't it be better to split it and to squash each of its parts in
the commit that removes the corresponding macros?

2. Would it make sense to ensure that the compatibility macros are
defined only outside of the runtime, to make sure we do not ourselves
restart using them?

You mayalso want to add a comment sayiing at which version the macro got
deprecated, so that it's easier to decide, later, that time for removal
has arrived.

In reference to what you wrote in the commit message: "If this list
keeps growing, it could be moved to a separate file.": it might be more
useful to have this comment in the code itself rather than in the commit
message. Also, after all, why not introducing the file right now and so
it would be its inclusiob which could be protected by ifdef not
CAML_INTERNAL.

And, out of curiosity, re:"add missing quotes in branch", you mention
that using autoconf-2.70 gives an error but I thought we were usignit
already so how is it possible that this got unnoticed?

**Comment by shindere** [comment]
David Allsopp (2024/05/15 04:19 -0700):
> I like every aspect of this PR, thank you 🙂 Adding a green tick, but
> I'll let @shindere have a chance to commit before merging.

Many thanks for having reviewed AND given me the opportunity to do
likewise! Would have been a pity to miss such a nice work.

**Comment by MisterDA** [comment]
> 1. Wouldn't it be better to split it and to squash each of its parts in the commit that removes the corresponding macros? 

Will do.

> 2. […] Also, after all, why not introducing the file right now and so it would be its inclusion which could be protected by `#ifndef CAML_INTERNAL`.

Great idea, I'll re-introduce `compatibility.h`.

> And, out of curiosity, re:"add missing quotes in branch", you mention that using autoconf-2.70 gives an error but I thought we were using it already so how is it possible that this got unnoticed?

Without the `[ … ]` quotes, the code is brittle and sensitive to the context. When I removed other occurrences of `AC_CHECK_DECL` which occurred before this bug, Autoconf stopped expanding the macro before the bug, but expanded it at its sole occurrence inside the expansion of the `AS_IF` macro, which led to a wrong quoting, wrong nesting of if blocks and parentheses.

**Comment by shindere** [comment]
Antonin Décimo (2024/05/15 22:10 -0700):
> > 1. Wouldn't it be better to split it and to squash each of its parts in the commit that removes the corresponding macros? 
> 
> Will do.

Cool, thanks!

> > 2. […] Also, after all, why not introducing the file right now and so it would be its inclusion which could be protected by `#ifndef CAML_INTERNAL`.
> 
> Great idea, I'll re-introduce `compatibility.h`.

Thanks! That, or `legacy.h` as matches your taste best.

> > And, out of curiosity, re:"add missing quotes in branch", you mention that using autoconf-2.70 gives an error but I thought we were using it already so how is it possible that this got unnoticed?
> 
> Without the `[ … ]` quotes, the code is brittle and sensitive to the
> context. When I removed other occurrences of `AC_CHECK_DECL` which
> occurred before this bug, Autoconf stopped expanding the macro before
> the bug, but expanded it at its sole occurrence inside the expansion
> of the `AS_IF` macro, which led to a wrong quoting, wrong nesting of
> if blocks and parentheses.

Ah okay so it's because it was fragile, not because it was strictly
wrong, AIUI. Thanks for the explanation, it's enlightening!

**Comment by MisterDA** [comment]
Rebased. I went with `conf_compat.h`.

File: runtime/caml/config_compat.h
Comment (dra27) [inline runtime/caml/config_compat.h]: We can't co-opt `CAML_INTERNALS` this way, I'm afraid. Defining `CAML_INTERNALS` means you've opted _in_ to the internals; its negation does not mean that we're building OCaml itself!

FWIW I wouldn't worry about guarding this against internal usage, so I'd just unconditionally define them, as before.

**Comment by MisterDA** [comment]
Rebased for the MSVC CI fix.

**Comment by shindere** [comment]
David Allsopp (2024/05/16 00:22 -0700):
> We can't co-opt `CAML_INTERNALS` this way, I'm afraid. Defining `CAML_INTERNALS` means you've opted _in_ to the internals; its negation does not mean that we're building OCaml itself!

Ah, makes sense indeed.

Apologies for the misleading suggestion.

> FWIW I wouldn't worry about guarding this against internal usage, so
> I'd just unconditionally define them, as before.

OK, I guess we could always have a CI test to ensure that none of the
macros defined in the compatibility header is actually used by the rest
of the system, if we wanted.

**Comment by shindere** [comment]
Sorry, I'm getting confused: didn't we agree that "Keep former
Autoconf-defined macros for compatibility" would be split and melt into
the previous commits?

If I may nitpick about the name of the compatibility header: do we
really care that the macros come from `config`? Can't there be macros
from other places that we maywant to deprecate and whose definition
would move to a compatibility header during the deprecation period?

**Comment by MisterDA** [comment]
> Sorry, I'm getting confused: didn't we agree that "Keep former Autoconf-defined macros for compatibility" would be split and melt into the previous commits? 

I split the commit in two: we already had a definition of `HAS_STDINT_H` that I wanted to deprecate, so this went into the first commit, and served as introduction for the compatibility header. The code redefining `HAS_NANOSECOND_STAT` was merged in the other commit. I think this is rather clean.

> If I may nitpick about the name of the compatibility header: do we really care that the macros come from `config`? Can't there be macros from other places that we may want to deprecate and whose definition would move to a compatibility header during the deprecation period?

We have the example of the `compatibility.h` header from the 4.x branch for functions that were not namespaced. I was only concerned about macros related to configure, and I can't predict the future, which is why I went for a name indicating "config". It should be relatively harmless to use this header for whatever, so yes, I can go with `compatibility.h` and see what happens.

**Comment by shindere** [comment]
Antonin Décimo (2024/05/17 07:06 -0700):
OK works for me for the commits and yes please lets rename the header to
the more general! :)

**Comment by MisterDA** [comment]
> please lets rename the header to the more general! :)

Done.

**Review by shindere — APPROVED** [review]

**Pull request merged**
